registry/auth/basic/htpasswd.go

```diff
@@ -0,0 +1,150 @@
+package basic
+
+import (
+	"bufio"
+	"crypto/sha1"
+	"encoding/base64"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/docker/distribution/context"
+	"golang.org/x/crypto/bcrypt"
+)
+
+// htpasswd holds a path to a system .htpasswd file and the machinery to parse it.
+type htpasswd struct {
+	path string
+}
+
+// authType represents a particular hash function used in the htpasswd file.
+type authType int
+
+const (
+	authTypePlainText authType = iota // Plain-text password storage (htpasswd -p)
+	authTypeSHA1                      // sha hashed password storage (htpasswd -s)
+	authTypeApacheMD5                 // apr iterated md5 hashing (htpasswd -m)
+	authTypeBCrypt                    // BCrypt adapative password hashing (htpasswd -B)
+	authTypeCrypt                     // System crypt() hashes.  (htpasswd -d)
+)
+
+var bcryptPrefixRegexp = regexp.MustCompile(`^\$2[ab]?y\$`)
+
+// detectAuthCredentialType inspects the credential and resolves the encryption scheme.
+func detectAuthCredentialType(cred string) authType {
+	if strings.HasPrefix(cred, "{SHA}") {
+		return authTypeSHA1
+	}
+	if strings.HasPrefix(cred, "$apr1$") {
+		return authTypeApacheMD5
+	}
+	if bcryptPrefixRegexp.MatchString(cred) {
+		return authTypeBCrypt
+	}
+	// There's just not a great way to distinguish between these next two...
+	if len(cred) == 13 {
+		return authTypeCrypt
+	}
+	return authTypePlainText
+}
+
+// String Returns a text representation of the AuthType
+func (at authType) String() string {
+	switch at {
+	case authTypePlainText:
+		return "plaintext"
+	case authTypeSHA1:
+		return "sha1"
+	case authTypeApacheMD5:
+		return "md5"
+	case authTypeBCrypt:
+		return "bcrypt"
+	case authTypeCrypt:
+		return "system crypt"
+	}
+	return "unknown"
+}
+
+// NewHTPasswd Create a new HTPasswd with the given path to .htpasswd file.
+func newHTPasswd(htpath string) *htpasswd {
+	return &htpasswd{path: htpath}
+}
+
+// AuthenticateUser checks a given user:password credential against the
+// receiving HTPasswd's file. If the check passes, nil is returned. Note that
+// this parses the htpasswd file on each request so ensure that updates are
+// available.
+func (htpasswd *htpasswd) authenticateUser(ctx context.Context, username string, password string) error {
+	// Open the file.
+	in, err := os.Open(htpasswd.path)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+
+	for _, entry := range parseHTPasswd(ctx, in) {
```

Comment:
> we could build up a map?

In this case we have to track if file is changed.
Even having the map, it will take different times for existent and non-existent  users due to a password verification.
So we might assume that user exists if it takes more time... or GC comes :).